### PR TITLE
Fix assembly loading for NUnit with MTP

### LINKED ISSUE #335: Support NUnit with Microsoft.Testing.Platform v2
### Background and motivation

Upcoming changes in .NET10, NUnit3TestAdapter and Microsoft.Testing.Platform require changes in the discovery of NUnit test framework assembly.

Boundary conditions.
* Affects some test project configurations which use NUnit3TestAdapter v6 (to be release in the next days)
* Build with .NET 10 >= RC2 (Target framework might also be .NET8)

It can happen that case nunit.framework is loaded more than once, and our current test framework assembly discovery selects the wrong assembly (from the wrong `AssemblyLoadContext`).

For a longer version see https://github.com/nunit/nunit3-vs-adapter/issues/1348

In the end the fix should be easy, we must use 

```C#
    private static IEnumerable<Assembly> GetAssemblies()
    {
#if NET6_0_OR_GREATER
        // In some constellations a test framework assembly might have been loaded more than once. Make sure we get the correct one.
        return System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly()).Assemblies;
#else
        return AppDomain.CurrentDomain.GetAssemblies();
#endif
    }
```
in `LateBoundTestFramework.FindExceptionAssembly`.

The main point is, that we must make sure, that no existing constellations break.

I'll tackle this as soon as NUnit3TestAdapter v6 and .NET10 are released.

### Alternative Concerns

_No response_

### Are you willing help with a pull-request?

Yes, please assign this issue to me.

### PR DESCRIPTION
**Pull request by cbersch**
Fix #335

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [ ] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [ ] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/main/Tests/AwesomeAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/main/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://awesomeassertions.org/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/main/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/main/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/main/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://awesomeassertions.org/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

## Legal checklist

* [x] This work is entirely original, it is not derived from any existing code incompatible with the Apache 2.0 License, like FluentAssertions.

### CODE REVIEW

File: Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs [hunk 1]
```diff
@@ -23,20 +30,37 @@ public bool IsAvailable
...
-    private Assembly FindExceptionAssembly()
+    protected internal abstract string AssemblyName { get; }
+
+    protected abstract string ExceptionFullName { get; }
+
+    private static IEnumerable<Assembly> GetAssemblies()
     {
-        var assembly = Array.Find(AppDomain.CurrentDomain.GetAssemblies(), a => a.GetName().Name == AssemblyName);
+#if NET6_0_OR_GREATER
+        // In some constellations a test framework assembly might have been loaded more than once. Make sure we get the correct one.
```
Comment (lg2de) [inline Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs]: "Make sure we get the correct one" does not really help to understand, what makes the difference here.
Could you please add some information what is different when using `AssemblyLoadContext` instead of `AppDomain`?

File: Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    <TargetFrameworks>net8.0;net10.0</TargetFrameworks>
+    <RootNamespace>NUnit4.Mtp.Specs</RootNamespace>
+    <AssemblyName>NUnit4.Mtp.Specs</AssemblyName>
+    <EnableNUnitRunner>true</EnableNUnitRunner>
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\Src\AwesomeAssertions\AwesomeAssertions.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="coverlet.collector" Version="6.0.4" PrivateAssets="all">
```
Comment (lg2de) [inline Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj]: Do we really need "coverlet" here?
Some of these test projects have such reference, others not.

File: Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs (hunk 1 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs]: [AppDomain.GetAssemblies](https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.getassemblies?view=net-10.0): Gets the assemblies that have been loaded into the execution context of this application domain.
And in the case of NUnit4.Mtp.Specs this returns nunit.framework twice, with the first one being the wrong assembly, which is a different one than used for running the tests.

So we are looking for the nunit.framework assembly which is used during test execution, which we get through
[AssemblyLoadContext](https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/understanding-assemblyloadcontext).

Comment (IT-VBFK) [reply Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs]: Maybe worth a comment here?

```suggestion
        // In some constellations a test framework assembly might have been loaded more than once. Make sure we get the correct one:
        // AppDomain.GetAssemblies: Gets the assemblies that have been loaded into 
        // the execution context of this application domain.
        // And in the case of NUnit4.Mtp.Specs this returns nunit.framework twice, with the first one 
        // being the wrong assembly, which is a different one than used for running the tests.
        
        // So we are looking for the nunit.framework assembly which is used during test execution,
        // which we get through AssemblyLoadContext.
```

File: Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj (hunk 2 above)
Comment (cbersch) [reply Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj]: Probably I should reference ` Microsoft.Testing.Extensions.CodeCoverage` like the TUnit.Specs.csproj does.

File: Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs (hunk 1 above)
Comment (lg2de) [reply Src/AwesomeAssertions/Execution/LateBoundTestFramework.cs]: This is exactly what I meant to ask for!

**Review by IT-VBFK — APPROVED** [review]

**Comment by cbersch** [comment]
I just saw, that I forgot to add the new project to the build.

File: Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj (hunk 2 above)
Comment (cbersch) [reply Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj]: Did this, and also fixed the report generation for Microsoft.Testing.Platform specs.
Now coveralls goes crazy

**Comment by cbersch** [comment]
> I just saw, that I forgot to add the new project to the build.

Fixed

File: Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj (hunk 2 above)
Comment (cbersch) [reply Tests/TestFrameworks/MicrosoftTestingPlatform/NUnit4.Mtp.Specs/NUnit4.Mtp.Specs.csproj]: I just removed the coverage changes in `Build.cs`. So, the coverage from running the `MicrosoftTestingPlatform/*Specs.csproj` isn't considered, as on main. I added the same package references as the TUnit.Specs has.
This looks to me like yet another MTP bug :/ https://github.com/microsoft/testfx/issues/7082
